generate_charts.py

```diff
@@ -0,0 +1,206 @@
+    """Generates an empty chart figure message.
+    This is used to replace the chart object
+    when no chart is available.
+
+    Args:
+        message (str): The message to display in the center of the chart.
+
+    Returns:
+        go.Figure: A Plotly figure object containing the message.
+    """
+    fig = go.Figure()
+    fig.update_layout(
+        xaxis={"visible": False},
+        yaxis={"visible": False},
+        annotations=[
+            {
+                "text": message,
+                "xref": "paper",
+                "yref": "paper",
+                "showarrow": False,
+                "font": {"size": 28},
+            }
+        ],
+    )
+    fig.update_layout(
+        margin=dict(l=20, r=20, t=10, b=10),
+    )
+    return fig
+
+
+def generate_gantt_chart(
+    scenario: str = None, df: pd.DataFrame = None, y_axis: str = "Job", color: str = "Resource"
+) -> go.Figure:
+    """Generates a Gantt chart of the unscheduled tasks for the given scenario.
+
+    Args:
+        scenario (str): The name of the scenario; must be a key in SCENARIOS.
+        df (pd.DataFrame): A DataFrame containing the data to plot. If this is
+            not None, then the scenario argument will be ignored.
+        y_axis (str): The column to use for the y-axis of the Gantt chart.
+        color (str): The column to use for the color of the Gantt chart.
+
+    Returns:
+        go.Figure: A Plotly figure object.
+    """
+    if df is None:
+        filename = SCENARIOS[scenario]
+        if "json" in filename:
+            model_data.load_from_json(DATA_PATH.joinpath(filename), resource_names=RESOURCE_NAMES)
+        else:
+            model_data.load_from_file(DATA_PATH.joinpath(filename), resource_names=RESOURCE_NAMES)
+        df = get_minimum_task_times(model_data)
+    if y_axis == "Job":
+        if df["Job"].dtype == "object":
+            df["JobInt"] = df["Job"].str.replace("Job", "").astype(int)
+        else:
+            df["JobInt"] = df["Job"]
+        df = df.sort_values(by=["JobInt", color, "Start"])
+        df = df.drop(columns=["JobInt"])
+    else:
+        df = df.sort_values(by=[y_axis, color, "Start"])
+    df["delta"] = df.Finish - df.Start
+    df[color] = df[color].astype(str)
+    df[y_axis] = df[y_axis].astype(str)
+    num_items = len(df[color].unique())
+    colorscale = "Agsunset"
+    fig = px.timeline(
+        df,
+        x_start="Start",
+        x_end="Finish",
+        y=y_axis,
+        color=color,
+        color_discrete_sequence=px.colors.sample_colorscale(
+            colorscale, [n / (num_items - 1) for n in range(num_items)]
+        ),
+    )
+
+    for idx, _ in enumerate(fig.data):
+        resource = fig.data[idx].name
+        data_list = []
+        for job in fig.data[idx].y:
+            try:
+                data_list.append(
+                    df[(df[y_axis] == job) & (df[color] == resource)].delta.tolist()[0]
+                )
+            except:
+                continue
```

Comment:
Should catch a specific error here, not all.

Reply:
This whole for-loop seems a bit unnecessary actually. I think it just scales the x-values, which can probably be done in a better, less complicated way. I might be missing something though.

Reply:
updated: https://github.com/dwave-examples/job-shop-scheduling-cqm/pull/17/commits/5e41bd2e6de63f470bc2ac073b985861d81f7698